Reduces Unnecessary Duplicated Sales Invoice Generation

DanielWright opened the pull request:
An Avatax sales invoice doesn't need to be generated each time tax-adjustment is calculated for a single line-item or shipment.  Once per order is just fine, thankyouverymuch. This PR proposes a change of this behaviour by overriding `Spree::Tax::OrderAdjuster#adjust!`.  When all the tax-rates for the order's zone are Avatax-calculated, the entire order is adjusted by the `SalesInvoice.generate` call, a sales invoice is generated, and the item-by-item adjustment phase is short-circuited. Without any Avatax-calculated rates, the normal item-by-item adjustment phase proceeds as before.

As a result, `TaxRate#adjust` is once again a no-op when the tax-rate is Avatax-calculated.

Finally, a number of failing specs have been fixed.

Commit:
Reduces Unnecessary Duplicated Sales Invoice Generation

An Avatax sales invoice doesn't need to be generated each time tax-adjustment
is calculated for a single line-item or shipment.  Once per order is just fine,
thankyouverymuch. This change in behaviour is achieved by overriding
`Spree::Tax::OrderAdjuster#adjust!`.  When all the tax-rates for the order's
zone are Avatax-calculated, a sales invoice is generated, and the rest of the
operation is skipped. The normal adjustment implementation is followed in any other case.

Additionally, a sales-invoice is no longer generated in Avatax-calculated tax
rates. Calling `TaxRate#adjust` in these cases in a no-op.

Commit:
Fixes Spree::Order Specs

The fix introduced in 646544e to avoid committing non-existent sales invoices
introduced failures into the spec.

Commit:
Fixes Broken SpreeAvatax::ShortShipReturnInvoice Spec

This inadvertently reverts a change in
186a8d57b826fb1e19213ab51fa6bfc74ca9dccd. I cannot explain why changing the
number back has fixed the spec, any more than I could explain it last time...

Commit:
Fixes SpreeAvatax::ReturnInvoice Specs

Changes introduced in #10 broke the return-invoice specs. I've altered the
specs so they anticipate the necessary state to satisfy #10's new logic.

braidn commented on spec/models/spree/order_spec.rb:
```diff
@@ -65,7 +67,7 @@ def change_address
 
     it "generates a sales invoice" do
-      expect(SpreeAvatax::SalesInvoice).to receive(:generate).with(order)
+      expect(SpreeAvatax::SalesInvoice).to receive(:generate).with(order).exactly(:once)
```
💥 ! Reads like a 📖

braidn commented on spec/models/spree/tax/order_adjuster_spec.rb:
```diff
@@ -0,0 +1,68 @@
+      allow(item_adjuster_klass).to receive(:new).and_return(item_adjuster)
+      allow(item_adjuster).to receive(:adjust!)
+
+      stub_const("Spree::Tax::ItemAdjuster", item_adjuster_klass)
```
I just adore the 'transfer_nested_constants' method available on these stubs. The magic here in the implementation must be bonkers.

braidn commented on spec/models/spree/tax/order_adjuster_spec.rb:
```diff
@@ -0,0 +1,68 @@
+      allow(sales_invoice_klass).to receive(:generate)
+    end
+
+    context "with an order in an Avatax-calculated zone" do
```
Nice context grouping here. These read great!

braidn commented on app/models/spree/tax/order_adjuster_decorator.rb:
```diff
@@ -0,0 +1,14 @@
+  module Extensions
+    module OrderAdjuster
+      def adjust!
+        return super unless rates_for_order_zone(order).all?(&:avatax?)
```
I think this may be a problem right? This, if I understand, will grab all rates for the order. Then, it is asking if all of these rates respond true for avatax. Should this be `all` or `any`? I have been playing on the [PR app](https://glossier-pr-3151.herokuapp.com) and it doesn't seem to be generating an invoice/sales tax for an order sent to the Lafayette office.

braidn replied:
I also ran this code in the terminal against the mentioned order (id: 2841345) using the following:
```
tax = Spree::Tax::OrderAdjuster.new(order)
tax.send(:rates_for_order_zone, order).all?(&:avatax?)
  Spree::Calculator Load (1.3ms)  SELECT  "spree_calculators".* FROM "spree_calculators" WHERE "spree_calculators"."calculable_id" = $1 AND "spree_calculators"."calculable_type" = $2 LIMIT 1  [["calculable_id", 1], ["calculable_type", "Spree::TaxRate"]]
=> false

tax.send(:rates_for_order_zone, order).any?(&:avatax?)
  Spree::Calculator Load (1.6ms)  SELECT  "spree_calculators".* FROM "spree_calculators" WHERE "spree_calculators"."calculable_id" = $1 AND "spree_calculators"."calculable_type" = $2 LIMIT 1  [["calculable_id", 2], ["calculable_type", "Spree::TaxRate"]]
  Spree::Calculator Load (1.4ms)  SELECT  "spree_calculators".* FROM "spree_calculators" WHERE "spree_calculators"."calculable_id" = $1 AND "spree_calculators"."calculable_type" = $2 LIMIT 1  [["calculable_id", 3], ["calculable_type", "Spree::TaxRate"]]
=> true
```

DanielWright replied:
Hmm. I made the assumption that Avatax would be the only tax-rate covering a given order. Maybe that was a bad assumption?

braidn replied:
No, it's a legit assumption. No worries there. However, say we focus on just the Lafayette address, this will pull in North America along with Shopify since we have a Shopify popup above the office. This will return both tax rates. The Avatax powered North America one and the none Avatax powered Shopify one.

DanielWright replied:
Then I think you're right. Unfortunately, if I go with `any?`, then I have to go through the `ItemAdjuster`. I will make some adjustments and resubmit. :+1:

braidn reviewed (commented):
Perhaps the PR is oddly setup? Everything looks correctly setup but, I could be off.

Commit:
Handle Mixed Tax-Rate Zones

In cases where an order belongs to a zone taxed by one Avatax-calculated rate
and another rate, the Avatax sales-invoice would clobber the other tax
adjustments.

This issue is here resolved by constraining the behaviour of
`SalesShared.reset_tax_attributes`. It will now destroy only Avatax-based tax
adjustments, instead of all adjustment. It will now destroy only Avatax-based
tax adjustments, instead of all adjustments.

braidn commented:
@mimsterynyc would it be possible to you run a Q/A round here? We need to make sure that we are generating Tax properly (especially for areas where zones may overlap). This is likely NYC. Thank you.

braidn commented:
PR APP is over here: https://glossier-pr-3151.herokuapp.com/

braidn commented:
Looking good though. It's now calling out to Avatax. This is disabled on this app but, re-enabling it now.

mimsterynyc commented:
Mimi Middleton (123) [9:41 AM] 
Hi Daniel


[9:42] 
i have a question...questions really, about the Avatax sales invoice


Daniel Wright [9:42 AM] 
Shoot!


Mimi Middleton (123) [9:42 AM] 
i'm not sure i fully understand the issue, and i need to in order to test it lol


Daniel Wright [9:43 AM] 
I will give you what answers I can.


[9:43] 
Do you want to ask questions, or would you like a broad-strokes rundown?


Mimi Middleton (123) [9:43 AM] 
maybe a rundown first


Daniel Wright [9:43 AM] 
Sure.


[9:45] 
I'm not sure how much the nomenclature matters, but a "sales invoice" is basically a document at Avalara, which indicates what items were comprised in a sale, and what taxes are owed for those items. "Items" in this case can be, in Solidus parlance, line-items or shipments.


[9:45] 
(I"m commuting and have to get off the bus. To be continued..!)


Daniel Wright [9:55 AM] 
As far as Solidus is concerned, there are two "phases" the submission of this sales-invoice goes through: generating the invoice, and committing it.


[9:57] 
At a couple stages of the checkout, Solidus generates (or re-generates) the sales-invoice, to accommodate things like the shipping address and the delivery method. (Both of these things will change the amount or jurisdiction of the tax.)


[9:58] 
When the checkout is completed, the sales-invoice is committed. This tells Avalara that the invoice is final, and can't be changed. The details therein are the taxes that should be collected and disbursed to the appropriate jurisdictions.

[9:59] 
The implementation in production is less than ideal for one main reason: the generating/re-generating of the sales-invoice is performed once _for every line-item and shipment_ in the order.

[9:59] 
This is expensive and unnecessary computation.

[10:00] 
The PR that Braden asked you to QA changes this behaviour so that the generating/re-generating happens once per request.

[10:00] 
Does that clear things up?

Mimi Middleton (123) [10:00 AM] 
yes.

[10:01] 
how do you see that it is happens more than once  (generate/regenerate) currently?

[10:01] 
on a report?

Daniel Wright [10:02 AM] 
Not directly, no. Alex raised a flag in technology (https://glossier.slack.com/archives/G0S7Q3NGH/p1508435843000601) that the number or duration of requests to Avalara APIs had gone way up.

[10:03] 
I investigated my previous code contribution, and realized my mistake.

[10:04] 
As part of the PR, I wrote some RSpec tests in the Avatax extension to set the expectation that Avalara APIs would only be contacted once per request, and then adjusted the implementation accordingly.

Mimi Middleton (123) [10:06 AM] 
ok so i understand, but how will i physically see that your fix is functioning accordingly


[10:06] 
i'm sorry, but i'm still learning alol


[10:06] 
LOL


Daniel Wright [10:07 AM] 
That's an excellent point. I'm not sure you can.


[10:08] 
I had thought the avenue of QA you'd pursue is making sure the PR didn't break taxation in any of our jurisdictions, but it would be better if you could also check that we've seen the expected decrease in API calls.


[10:08] 
Let me check w/ Braden about whether we can hook up NewRelic to the review app.


Daniel Wright [10:14 AM] 
I've also checked in with the tech channel to see how we typically QA performance improvements.


[10:14] 
I'm also learning. :slightly_smiling_face:


Mimi Middleton (123) [10:15 AM] 
Yeah i was just talking to Alex and Braden about maybe being able to QA the decrease amount of calls in API-


[10:16] 
let me put this convo in Github and see if my team has some input too


Daniel Wright [10:16 AM] 
Cool


[10:16] 
One challenge might be to establish the "before" level. I guess we can use production?

mimsterynyc commented:
@braidn @aamyot - do you have any ideas/input?

braidn commented:
If this is still generating tax for you Mimi (when and where it should be), introducing this PR to production and monitoring the calls to Avatax should be 👍

braidn commented:
Moving this into production now.
